File: android-prototype/src/model/features.js
// Port of handgesture/features.py — per-frame landmark features.
// All values are pairwise distances normalized by hand scale (wrist ->
// middle-MCP), so they're invariant to position, rotation, mirroring and
// camera distance. 14 base features + 14 deltas = 28 model inputs.

const TIPS = [4, 8, 12, 16, 20];
const CURL_PAIRS = [[4, 2], [8, 5], [12, 9], [16, 13], [20, 17]];
const GAP_PAIRS = [[4, 8], [8, 12], [12, 16], [16, 20]];

export const FEAT_DIM = 14;

// landmarks: array of 21 {x, y} in any consistent space (pixels preferred)
export function frameFeatures(lm) {
  const d = (a, b) => Math.hypot(lm[a].x - lm[b].x, lm[a].y - lm[b].y);
  const scale = d(9, 0) + 1e-6;
  const f = new Array(FEAT_DIM);
  let i = 0;
  for (const t of TIPS) f[i++] = d(t, 0) / scale;
  for (const [a, b] of CURL_PAIRS) f[i++] = d(a, b) / scale;
  for (const [a, b] of GAP_PAIRS) f[i++] = d(a, b) / scale;
  return f;
}

// Maintains the delta half of the feature vector across a streaming session.
export class FeatureStream {
  constructor() { this.prev = null; }
  reset() { this.prev = null; }
  next(lm) {
    const f = frameFeatures(lm);
    const x = new Array(2 * FEAT_DIM);
    for (let i = 0; i < FEAT_DIM; i++) {
      x[i] = f[i];
      x[FEAT_DIM + i] = this.prev ? f[i] - this.prev[i] : 0;
    }
    this.prev = f;
    return x;
  }
}

File: handgesture/features.py
"""
Per-frame landmark features for the temporal early-exit model.

All features are pairwise distances normalized by hand scale, so they are
invariant to translation, rotation, mirroring and camera distance. That
keeps synthetic training sequences and real MediaPipe landmarks in nearly
the same domain — no silhouette rendering needed.

Layout (14 dims per frame):
  0-4   fingertip -> wrist distance   (openness), thumb/index/middle/ring/pinky
  5-9   fingertip -> own MCP distance (curl)
  10-13 adjacent fingertip gaps       (spread)

sequence_features() appends first differences -> 28 dims per frame.
"""

import numpy as np

FEAT_DIM = 14
SEQ_FEAT_DIM = 2 * FEAT_DIM

_TIPS = [4, 8, 12, 16, 20]
_CURL_PAIRS = [(4, 2), (8, 5), (12, 9), (16, 13), (20, 17)]
_GAP_PAIRS = [(4, 8), (8, 12), (12, 16), (16, 20)]


def frame_features(landmarks) -> np.ndarray:
    """(21, 2+) landmark array -> (14,) float32 feature vector."""
    p = np.asarray(landmarks, dtype=np.float64)[:, :2]
    scale = np.linalg.norm(p[9] - p[0]) + 1e-6   # wrist -> middle MCP

    def d(a, b):
        return np.linalg.norm(p[a] - p[b]) / scale

    f = [d(t, 0) for t in _TIPS]
    f += [d(a, b) for a, b in _CURL_PAIRS]
    f += [d(a, b) for a, b in _GAP_PAIRS]
    return np.asarray(f, dtype=np.float32)


def sequence_features(seq) -> np.ndarray:
    """(T, 21, 2+) landmark sequence -> (T, 28) features (values + deltas)."""
    feats = np.stack([frame_features(fr) for fr in seq])
    deltas = np.diff(feats, axis=0, prepend=feats[:1])
    return np.concatenate([feats, deltas], axis=1)
